### ISSUE TITLE
warp-ctc and torch.nn.CTCLoss

### ISSUE THREAD

**Issue by xiiiiiiii** [Holmeyoung/crnn-pytorch#14]
我想问一下warp-ctc是不是在直接是rnn的输出就好了？但是nn.ctcloss需要加一层softmax才能输入到ctc中，但是我看你的代码没有加softmax，这样是可以的吗？另外好像官网更新了ctcloss

**Comment by Holmeyoung** [comment]
你看到的那个是不是 `log_softmax` 呀，官网的例子贴一下先
```python
>>> T = 50      # Input sequence length
>>> C = 20      # Number of classes (excluding blank)
>>> N = 16      # Batch size
>>> S = 30      # Target sequence length of longest target in batch
>>> S_min = 10  # Minimum target length, for demonstration purposes
>>>
>>> # Initialize random batch of input vectors, for *size = (T,N,C)
>>> input = torch.randn(T, N, C).log_softmax(2).detach().requires_grad_()
>>>
>>> # Initialize random batch of targets (0 = blank, 1:C+1 = classes)
>>> target = torch.randint(low=1, high=C+1, size=(N, S), dtype=torch.long)
>>>
>>> input_lengths = torch.full(size=(N,), fill_value=T, dtype=torch.long)
>>> target_lengths = torch.randint(low=S_min, high=S, size=(N,), dtype=torch.long)
>>> ctc_loss = nn.CTCLoss()
>>> loss = ctc_loss(input, target, input_lengths, target_lengths)
>>> loss.backward()
```
取 `softmax` 再取 `log` 是改变了一下数据的分布，但是还是回到原数据上去了
```python
import torch
import torch.nn.functional as F

data = torch.FloatTensor([1.0,2.0,3.0,5.0])
softmax = F.softmax(data, dim=0)
print(softmax)

log_softmax = F.log_softmax(data, dim=0)
print(log_softmax)
```
输出
```bash
tensor([0.0152, 0.0414, 0.1125, 0.8310])
tensor([-4.1852, -3.1852, -2.1852, -0.1852])
```

**Comment by xiiiiiiii** [comment]
> 你看到的那个是不是 `log_softmax` 呀，官网的例子贴一下先
> 
> ```python
> >>> T = 50      # Input sequence length
> >>> C = 20      # Number of classes (excluding blank)
> >>> N = 16      # Batch size
> >>> S = 30      # Target sequence length of longest target in batch
> >>> S_min = 10  # Minimum target length, for demonstration purposes
> >>>
> >>> # Initialize random batch of input vectors, for *size = (T,N,C)
> >>> input = torch.randn(T, N, C).log_softmax(2).detach().requires_grad_()
> >>>
> >>> # Initialize random batch of targets (0 = blank, 1:C+1 = classes)
> >>> target = torch.randint(low=1, high=C+1, size=(N, S), dtype=torch.long)
> >>>
> >>> input_lengths = torch.full(size=(N,), fill_value=T, dtype=torch.long)
> >>> target_lengths = torch.randint(low=S_min, high=S, size=(N,), dtype=torch.long)
> >>> ctc_loss = nn.CTCLoss()
> >>> loss = ctc_loss(input, target, input_lengths, target_lengths)
> >>> loss.backward()
> ```
> 
> 取 `softmax` 再取 `log` 是改变了一下数据的分布，但是还是回到原数据上去了
> 
> ```python
> import torch
> import torch.nn.functional as F
> 
> data = torch.FloatTensor([1.0,2.0,3.0,5.0])
> softmax = F.softmax(data, dim=0)
> print(softmax)
> 
> log_softmax = F.log_softmax(data, dim=0)
> print(log_softmax)
> ```
> 
> 输出
> 
> ```shell
> tensor([0.0152, 0.0414, 0.1125, 0.8310])
> tensor([-4.1852, -3.1852, -2.1852, -0.1852])
> ```

是的，是log_softmax,
所以nn.CTCloss应该在网络的最后边添加log_softmax,比如
`out = self.classifier(out)
out = F.log_softmax(out, dim=2)
return out`
但是我看到你的代码中没有log_softmax,他取了个softmax再去log是回到原数据上了,但是他和原数据的值是不一样的，这样子是可以的吗？可以不加log_softmax吗？

**Comment by Holmeyoung** [comment]
可以不加呀，如果加了只是结果的偏移的话，在训练的时候就相当于学习有偏移的权重了～

**Comment by xiiiiiiii** [comment]
明白了，谢谢


 
---Original---

**Comment by Holmeyoung** [comment]
嗨，最新的代码里还是加进了 `log_softmax`，你之前说的是正解呢～～～感谢～～～

**Comment by ivoxx** [comment]
我之前没加log_softmax，感觉结果也可以啊。加不加对结果影响大吗？

**Comment by Holmeyoung** [comment]
我没对比测试。但是强制规范数据分布，想必会有影响的。不一定体现在精度上，收敛速度应该不一样。哈哈哈，我也是猜得。官方是推荐放在log_softmax后面的，就告诉你一哈～～～很感谢之前提醒！
